src/xopen/__init__.py

```diff
@@ -330,18 +343,23 @@ def __iter__(self):
     def __next__(self) -> AnyStr:
         return self._file.__next__()
 
-    def _raise_if_error(self, allow_sigterm: bool = False) -> None:
+    def _raise_if_error(self,
+                        allowed_error_code: Optional[int] = None,
+                        allowed_message: Optional[Pattern] = None) -> None:
         """
-        Raise IOError if process is not running anymore and the exit code is
-        nonzero. If allow_sigterm is set and a SIGTERM exit code is
-        encountered, no error is raised.
+        Raise OSError if process is not running anymore and the exit code is
+        nonzero. If allowed_error_code is set and the exit value of the process is
+        equal to the value of allowed_error_code do not raise an error. If allowed_message
+        is set and does not match the output from stderr, OSError is raised even if the
+        exit code matches allowed_error_code.
         """
         retcode = self.process.poll()
+        message = self._stderr.read().strip()
         if (
             retcode is not None and retcode != 0
-            and not (allow_sigterm and retcode == -signal.SIGTERM)
+            and not (allowed_error_code is not None and retcode == allowed_error_code)
+            and not (allowed_message is not None and re.search(allowed_message, message))
```

Comment:
I think the description in the docstring doesn’t match the code. 

This long list of conditions is getting hard to understand. One way to make it more readable is to negate the condition and to split it up into multiple ifs (De Morgan helps with this). That way, one could check the conditions under which we *don’t* need to raise an error one by one and return early. It would look sth. like this (note this is not what the final code should be, see below):
```python
if retcode is None:
    # process still running
    return
if retcode == 0:
    # process terminated successfully
    return
if retcode == allowed_error_code:
    # terminated with allowed error code
    return
if allowed_message is not None and allowed_message in message:
    # terminated with another exit code, but message is allowed
    return
...
raise OSError(...)
```
However, spelling this out, I think this is not what you intended because exit code 1 in this case would never raise an error. The docstring also doesn’t match the logic in the code.

Reply:
Ah, I think you can use the above as-is if you just don’t set an allowed error code for `pbzip2` (set it to `None`) *only* an allowed message.